### PR TITLE
Changes to linked variables

### LINKED ISSUE #693: "linked" variables that call functions of constants and/or variables
@bqpd, had this issue using the function that we worked on yesterday.  It seems that all the variables that I am calculating using `lambda` functions are treated as free variables in the model.  Shouldn't they be constants? Because it's saying I can get more of this because nothing is constraining me... 

Here are my variables that I am calculating via `lambda`` function

``` python
              T_atm = VectorVariable(NSeg, 'T_{atm}',
                                      lambda T_sl, L_atm, h: T_sl - L_atm*h, 'K',
                                      'Air temperature', args=[T_sl, L_atm, h])
              P_shaftmax = VectorVariable(NSeg, 'P_{shaft-max}',
                                      lambda P_shaftmaxMSL, h_ref, h:
                                      P_shaftmaxMSL*(1-0.5178*(h/h_ref)**0.92),
                                      'hp', 'Max shaft power at altitude',
                                      args=[P_shaftmaxMSL, h_ref, h])
```

In the output, they are both under the free vars section: (obviously incorrect values)

```
\vec{T_{atm}} : [ 1.43e+05  403       1.47e+05  303      ... ]  [K]         Air temperature
\vec{P_{shaft-max}} : [ 6.37e+07  0.838     1.15e+08  1.18     ... ]  [hp]        Max shaft power at altitude  
```

### PR DESCRIPTION
**Pull request by bqpd**
@whoburg, ready for review

Closes #693

### CODE REVIEW

**Comment by bqpd** [comment]
@cjk12, this makes the syntax for https://github.com/hoburg/gpkit/blob/ad1947e3a6442dbed3d76d9538ea4f2be41dd7c2/docs/source/examples/external_sp2.py a lot simpler

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
Failing on some GPs in model tests, which is odd because I barely touched GPs...

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
@galbramc, this is failing on `gas_simple` on macys and Windows7 cvxopt, which is odd: I barely touched GPs. Does macys potentially also have an old version of cvxopt?

**Comment by galbramc** [comment]
I only see the error in Windows 7. I don't see an error on macys...?

**Comment by bqpd** [comment]
Must have disappeared on that last retest

**Comment by galbramc** [comment]
I don't see a macys error on any build associated with this PR...

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by whoburg** [comment]
This was merged without checks passing and unsurprisingly broke master. Working on cleaning this up now... confused why this was merged without allowing tests to pass.

File: gpkit/nomials/variables.py [hunk 1]
```diff
@@ -164,6 +165,14 @@ def __new__(cls, shape, *args, **descr):
             vl[i] = Variable(**descr)
             vl[i].key.arraykey = arraykey
 
+        if descr.pop("newvariable", True):
+            from .. import MODELS, MODELNUMS, NAMEDVARS
```
Comment (whoburg) [inline gpkit/nomials/variables.py]: comparing this to the nearly-identical code in `Variable`, this one is missing

```python
            if MODELS and MODELNUMS:
                NAMEDVARS[tuple(MODELS), tuple(MODELNUMS)].append(self)
```
Should that be added, or should `NAMEDVARS` be removed from the import?

Comment (bqpd) [inline gpkit/nomials/variables.py]: The latter was the right thing to do! `NAMEDVARS` gets added to during Variable creation, which happens during VectorVariable creation...

File: gpkit/constraints/prog_factories.py [hunk 2]
```diff
@@ -19,7 +19,7 @@
 
 def _progify_fctry(program, return_attr=None):
     "Generates function that returns a program() and optionally an attribute."
-    def programify(self, verbosity=1, substitutions=None, **kwargs):
+    def programify(self, verbosity=1, constants=None, **kwargs):
```
Comment (whoburg) [inline gpkit/constraints/prog_factories.py]: is constants a substitutions dict? or something else

File: gpkit/constraints/signomial_program.py [hunk 3]
```diff
@@ -151,7 +154,18 @@ def gp(self, x0=None, verbosity=1):
...
+        self.__add_externalfns_maybe()
+        if self.externalfn_vars:
+            gp_constrs.extend([v.key.externalfn(v, x0)
+                               for v in self.externalfn_vars])
         gp = GeometricProgram(self.cost, gp_constrs,
                               self.substitutions, verbosity=verbosity)
         gp.x0 = x0  # NOTE: SIDE EFFECTS
         return gp
+
+    def __add_externalfns_maybe(self):
```
Comment (whoburg) [inline gpkit/constraints/signomial_program.py]: with that name I'm dying to see a docstring

File: gpkit/constraints/signomial_program.py [hunk 4]
```diff
@@ -48,6 +48,9 @@ def __init__(self, cost, constraints, substitutions=None, verbosity=1):
     (z >= s) and then minimizing that dummy variable.""")
         CostedConstraintSet.__init__(self, cost, constraints, substitutions)
         try:
+            self.__add_externalfns_maybe()
+            if self.externalfn_vars:
+                raise InvalidGPConstraint
```
Comment (whoburg) [inline gpkit/constraints/signomial_program.py]: informative error message?

File: gpkit/tests/t_constraints.py [hunk 5]
```diff
@@ -33,6 +33,14 @@ def test_bad_elements(self):
         with self.assertRaises(ValueError):
             _ = Model(x, [v >= ["A", "B"]])
 
+    def test_evalfn(self):
+        x = Variable("x")
+        x2 = Variable("x^2", evalfn=lambda solv: solv[x]**2)
+        m = Model(x, [x >= 2])
+        m.unique_varkeys = set([x2.key])
```
Comment (whoburg) [inline gpkit/tests/t_constraints.py]: mutating `unique_varkeys` seems odd here, wouldn't this happen automatically with a setup method? Should we test that?

File: gpkit/tests/t_sub.py [hunk 6]
```diff
@@ -210,13 +210,12 @@ def test_vector_sweep(self):
...
 
     def test_linked_sweep(self):
-        def night_hrs(day_hrs):
-            "twenty four minus day hours"
-            return 24 - day_hrs
-        t_day = Variable("t_{day}", "hours")
-        t_night = Variable("t_{night}", night_hrs, "hours", args=[t_day])
         x = Variable("x", "hours")
+        t_day = Variable("t_{day}", 12, "hours")
+        t_night = Variable("t_{night}", lambda c: 24 - c[t_day.key], "hours")
```
Comment (whoburg) [inline gpkit/tests/t_sub.py]: don't really need the `.key`, right?

File: gpkit/tools/tools.py [hunk 7]
```diff
@@ -141,7 +141,7 @@ def composite_objective(*objectives, **kwargs):
         descr = dict(w.descr)
         del descr["value"]
         descr["name"] = "v_{CO}"
-        w_s.append(Variable(value=('sweep', lambda x: 1-x), args=[w], **descr))
+        w_s.append(Variable(value=lambda const: 1-const[w.key], **descr))  # pylint: disable=cell-var-from-loop
```
Comment (whoburg) [inline gpkit/tools/tools.py]: I think pylint might be right here -- are you sure the values of `w` that end up in all these lambdas aren't the same after loop closure?

```python
In [12]: funs = [lambda x: x**i for i in range(4)]

In [13]: [f(2) for f in funs]
Out[13]: [8, 8, 8, 8]
```

Comment (bqpd) [inline gpkit/tools/tools.py]: I don't think it would run if that were the case.

File: gpkit/tests/t_constraints.py (hunk 5 above)
Comment (bqpd) [inline gpkit/tests/t_constraints.py]: I think this serves, but it could be rewritten.

File: gpkit/constraints/signomial_program.py (hunk 3 above)
Comment (bqpd) [inline gpkit/constraints/signomial_program.py]: Ahaha

File: gpkit/constraints/signomial_program.py (hunk 4 above)
Comment (bqpd) [inline gpkit/constraints/signomial_program.py]: Will add!

File: gpkit/constraints/prog_factories.py (hunk 2 above)
Comment (bqpd) [inline gpkit/constraints/prog_factories.py]: Yup yup

File: docs/source/examples/external_sp2.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+"Can be found in gpkit/docs/source/examples/external_sp.py"
+import numpy as np
+from gpkit import Variable, Model
+
+x = Variable("x")
+
+
+def y_ext(self, x0):
```
Comment (bqpd) [inline docs/source/examples/external_sp2.py]: I don't know! It's not anti-pythonic to do this, but is it confusing? I wanted to make sure people weren't just sticking arbitrary constraints into these. :p

File: docs/source/examples/external_sp2.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
+"Can be found in gpkit/docs/source/examples/external_sp.py"
```
Comment (bqpd) [inline docs/source/examples/external_sp2.py]: Will fix!

**Pull request merged**